### ISSUE TITLE
research alternatives to clearing all allocated memory

### ISSUE THREAD

**Issue by LADSoft** [LADSoft/OrangeC#479]
it seems like currently the biggest time draw in compiles comes from the need to zero all memory that is allocated.   Zeroing memory is relatively expensive because of the huge amounts of memory the compiler uses.   Some redesign may alleviate this, or it may be possible to come up with a processor-specific version of bzero that uses something like SSE or write-through-caching to speed up the zeroing operation.

**Comment by chuggafan** [comment]
I would like to add an addendum to this: Researching ways of removing large chunks of the allocation we do at all, the compiler has memory problems outright and changing data structures and removing excessive allocations is something that is necessary for the best outcome.

Our own compiler side? If we support SSE we can change BZero to be a C file, implement the SSE instrinsics, and operate on blocks at a time, but that still doesn't help us in the end because we should be minimizing the amount that we need to zero in the first place.

**Comment by LADSoft** [comment]
probably a good idea, however, what you propose is a huge task (just a part of it would be a complete overhaul of how templates are done which I agree is necessary but it is still huge).   Given that we already have a bunch of huge tasks ahead of that it may be advisable to just punt and see what we can do about zeroing memory in the short term...

**Comment by LADSoft** [comment]
I thought about this some more and am now thinking, maybe this is important enough that we delay everything else to work on it (after #473 of course).   What do you think?

**Comment by chuggafan** [comment]
I don't know, the changing of the data formats that we're sending between the compiler sections means that we definitely do #473 before this so that everything makes sense there, then we can change the data formats in one place and standardize everything there.

This specific project will be a VERY long process because of how allocations are working here. If we allocate and zero large blocks before use it might be better? Allocating once > allocating 30 times in terms of speed and zeroing once > zeroing 30 times. I suspect doing this (properly) will take incrementally months so I'd like to avoid putting what's already being worked on for the backburner for 6+ months. I would appreciate finishing the immediate task at hand first then working on speeding up the compiler in a separate pass.

**Comment by LADSoft** [comment]
cool.   I agree this is a very long task.   I've been thinking for a while that one of the things that needs doing is to modernize memory allocations and get rid of the existing allocator altogether (in favor of using new/delete everywhere).

FWIW it does already allocate large blocks (128K).   I've experimented with various things including win32 heaps (with HEAP_ZERO flag) and calloc and malloc/memset...   the zeroing gets it every time though sigh.

**Comment by LADSoft** [comment]
I am thinking there may be some low-hanging fruit on this issue.    Gonna work on getting the builds to work again before I research whether there is any truth to that though.

**Comment by LADSoft** [comment]
I did some research and while there are a lot of structures using up various amounts of space and so there is a lot of room for improvement, two of the biggest culprits seem to be the caching done for template bodies and the fact that the template handling copies a lot of 'symbol' structures around...   the symbol structures are huge but sparsely used in this context and the caching is taking up a lot of space as well.    Probably going to attempt to address these two issues over the next couple of weeks because I don't think the appveyor build is going to complete on #297 until I decide which way I want to go with speeding up the build process...    the alternative being to rework the builds to build less times and do some comparisons on the binaries for proof of validity.    Which possibly I ought to do anyway I guess...

**Comment by chuggafan** [comment]
Some other stuff that might help that I am currently working on is reorganizing structs so that they are more able to be "packed", while this might affect us if we use any UB tricks it shouldn't really do anything else, as it's just moving stuff around without changing types. If we used a packed attribute we might be able to squeeze even more bytes out but letting it be aligned to say, a cache size or less is preferable, since it also helps with making zeroing memory take less time. It will also help us while we have allocated memory as the structs won't take up as much space so we use less memory overall (though removing allocations is preferable).

**Comment by LADSoft** [comment]
are you trying to maintain alignment as well?

**Comment by LADSoft** [comment]
that does sound like a good idea though, i trimmed off quite a bit of time by simplying reducing memory allocations :)

**Comment by chuggafan** [comment]
Alignment would probably be effected, but if we can make it aligned along cache lines if they aren't something that can be when we're doing large arrays or things of the sort it may be faster than if they aren't aligned along cache values.

Main thing here is that some structs that we have literally take up 512 bytes of space (_symbody from SYMBOL is this one) when we really should be using less, and getting something from 48 bytes to 32 does have an effect. Just by removing the total number of bytes we need to allocate we save system memory overall and, in some cases, speed. Simply rearranging structures or choosing something better and reworking based around that can save allocation time and processing time.

**Comment by LADSoft** [comment]
yeah I had figured out _symbody was way too big...   that was why I made it (recently) so I could have smaller symbols when working with template arguments...    because template arguments are a large number of the allocations being done.    It helped somewhat...    something more useful though was lower the number of allocations GetClassTemplate() does when trying to choose a template from a bunch of specializations...   still uses way too much memory though sigh.

Another big memory hog is the whole 'deferred compile' mechanism.   It caches zillions of LEXEME structures and uses a large amount of memory...    that is what I was going to work on next as it seems fairly straightforward and might have a big payoff.    We'll see whether we need to do it immediately though based on how much time getting rid of OCIDE buys.

unfortunately at some point symbody became a catch-all for anything that could be maintained for any kind of symbol and a lot of the stuff in there is useless for a large number of symbols...   with some work there could be large decreases in size there but it would take a bunch of research to do it right...
